File: algorithms/[1432]改变一个整数能得到的最大差值/solution.js
/**
 * @param {number} num
 * @return {number}
 */
const maxDiff = function(num) {
  return getMaxNum(num) - getMinNum(num);
};

const getMaxNum = function(num) {
  const array = num.toString().split('');

  const x = array.find(v => v !== '9') || '9';
  const y = '9';

  const result = array.map(v => v === x ? y : v).join('');

  return Number(result);
}

const getMinNum = function(num) {
  const array = num.toString().split('');

  let x = array[0];
  let y = '1';

  if (array[0] === '1') {
    x = array.find((v, i) => i > 0 && v !== '0' && v !== '1') || '0';
    y = '0';
  }

  const result = array.map(v => v === x ? y : v).join('');

  return Number(result);
}

File: algorithms/[1432]改变一个整数能得到的最大差值/solution.py
# topics = ["字符串", "贪心算法"]


class Solution:
    def maxDiff(self, num: int) -> int:
        """
        Greedy
        time O(log(num)), space O(log(num))
        """
        return self.get_max_num(num) - self.get_min_num(num)

    def get_max_num(self, num: int) -> int:
        nums = list(str(num))
        x = None

        for i, n in enumerate(nums):
            if n == x:
                nums[i] = '9'
            elif not x and n != '9':
                x = n
                nums[i] = '9'

        return int(''.join(nums))

    def get_min_num(self, num: int) -> int:
        nums = list(str(num))
        x = None
        r = ''

        for i, n in enumerate(nums):
            if n == x:
                nums[i] = r
            elif not x and i == 0 and n != '1':
                x = n
                r = '1'
                nums[i] = r
            elif not x and i > 0 and n != '1' and n != '0':
                x = n
                r = '0'
                nums[i] = r

        return int(''.join(nums))
